Retry resolving TCP address in agnhost/guestbook

hasheddan opened the pull request:
Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:

Currently the guestbook application will fail if unable to resolve TCP address on first attempt. If pod networking is not setup when the application starts then it will be unable to resolve, leading to frequent failures. This moves the address resolution into the retry block so it will try again if unsuccessful on first attempt.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref #89262 

**Special notes for your reviewer**:

Not sure if this requires the `agnhost` version bump, but I optimistically went ahead and updated.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

hasheddan committed:
Retry resolving TCP address in agnhost/guestbook

Currently the guestbook application will fail if unable
to resolve TCP address on first attempt. If pod networking
is not setup when the application starts then it will be
unable to resolve, leading to frequent failures. This moves
the address resolution into the retry block so it will try
again if unsuccessful on first attempt.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan commented:
/assign @claudiubelu

hasheddan commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

aojea commented on test/images/agnhost/guestbook/guestbook.go:
```diff
@@ -77,16 +77,18 @@ func registerNode(registerTo, port string) {
 	for time.Since(start) < timeout {
+		_, err := net.ResolveTCPAddr("tcp", hostPort)
+		if err != nil {
+			log.Printf("--slaveof param and/or --backend-port param are invalid. %v. Retrying in 1 second.", err)
```
I think that the message should reflect better the error, maybe `unable to resolve %s --slaveof param and/or --backend-port param with error %v. Retrying in second`?

mkumatag replied:
I feel its better to print the value of a second than hardcoded value.

hasheddan replied:
Will make both of these updates :+1:

hasheddan committed:
Improve error messages in agnhost/guestbook

This updates the error messages when registering a
node to be more explicit about what error occurred
and how long it will wait to retry.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan commented:
@aojea @mkumatag PTAL :)

aojea commented:
/lgtm

hasheddan commented:
/test pull-kubernetes-e2e-kind

claudiubelu commented on test/images/agnhost/guestbook/guestbook.go:
```diff
@@ -77,19 +77,21 @@ func registerNode(registerTo, port string) {
-		return
-	}
 
 	request := fmt.Sprintf("register?host=%s", getIP(hostPort).String())
```
It can still fail here, there's a fatal in ``getIP``, which causes it to fail instantly without any retries.

hasheddan replied:
Thanks for pointing that out! Will refactor

hasheddan committed:
Move getting IP address into retry loop for agnhost/guestbook

Removes the fatal error from getIP and moves it to
retry loop so that application will not immediately
crash on failure.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan commented:
@claudiubelu would you mind circling back for another review here?

hasheddan commented:
/retest

hasheddan commented:
/test pull-kubernetes-integration

hasheddan commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

hasheddan commented:
/test pull-kubernetes-e2e-kind-ipv6

hasheddan commented:
/retest

hasheddan commented:
/test pull-kubernetes-e2e-kind

mkumatag commented:
/lgtm

claudiubelu commented:
Cool, looks better now. I've ran the guestbook test a couple of times with these changes, it passed consistently.

/approve

hasheddan commented:
/test pull-kubernetes-e2e-kind

hasheddan commented:
/test pull-kubernetes-e2e-gce-100-performance

hasheddan commented:
/test pull-kubernetes-integration

hasheddan commented:
/test pull-kubernetes-e2e-kind

hasheddan commented:
/test pull-kubernetes-verify

hasheddan commented:
/test pull-kubernetes-e2e-kind

hasheddan commented:
/test pull-kubernetes-integration

hasheddan commented:
/test pull-kubernetes-e2e-kind

hasheddan commented:
/test pull-kubernetes-integration
